lxd/device/nic_ovn.go

```diff
@@ -441,6 +442,65 @@ func (d *nicOVN) Start() (*deviceConfig.RunConfig, error) {
 				}
 			}
 
+			integrationBridgeNICName = vfRepresentor
+			peerName = vfDev
+		} else if d.config["acceleration"] == "vdpa" {
+			ovs := openvswitch.NewOVS()
+			if !ovs.HardwareOffloadingEnabled() {
+				return nil, fmt.Errorf("SR-IOV acceleration requires hardware offloading be enabled in OVS")
+			}
+
+			err := util.LoadModule("vdpa")
+			if err != nil {
+				return nil, fmt.Errorf("Error loading %q module: %w", "vdpa", err)
+			}
+
+			// If VM, then try and load the vhost_vdpa module first.
+			if d.inst.Type() == instancetype.VM {
+				err = util.LoadModule("vhost_vdpa")
+				if err != nil {
+					return nil, fmt.Errorf("Error loading %q module: %w", "vhost_vdpa", err)
+				}
+			}
+
+			integrationBridge := d.state.GlobalConfig.NetworkOVNIntegrationBridge()
+
+			// Find free VF exclusively.
+			network.SRIOVVirtualFunctionMutex.Lock()
+			vfParent, vfRepresentor, vfDev, vfID, err := network.SRIOVFindFreeVFAndRepresentor(d.state, integrationBridge)
+			if err != nil {
+				network.SRIOVVirtualFunctionMutex.Unlock()
+				return nil, fmt.Errorf("Failed finding a suitable free virtual function on %q: %w", integrationBridge, err)
+			}
+
+			// Claim the SR-IOV virtual function (VF) on the parent (PF) and get the PCI information.
+			vfPCIDev, pciIOMMUGroup, err = networkSRIOVSetupVF(d.deviceCommon, vfParent, vfDev, vfID, false, saveData)
+			if err != nil {
+				network.SRIOVVirtualFunctionMutex.Unlock()
+				return nil, err
+			}
+
+			revert.Add(func() {
+				_ = networkSRIOVRestoreVF(d.deviceCommon, false, saveData)
+			})
+
+			// Create the vDPA management device
+			vDPADevice, err = ip.AddVDPADevice(vfPCIDev.SlotName, saveData)
+			if err != nil {
+				network.SRIOVVirtualFunctionMutex.Unlock()
+				return nil, err
+			}
+
+			network.SRIOVVirtualFunctionMutex.Unlock()
+
+			// Setup the guest network interface.
+			if d.inst.Type() == instancetype.Container {
```

Comment:
@gabrielmougard @stgraber `acceleration=vdpa` cannot be used with containers right? So should be failing earlier during 2nd stage validation (when d.inst != nil) if the instance is a container. Or should we silently fallback to doing `acceleration=sriov` mode like it looks like its doing here.

Reply:
@stgraber has confirmed:

> I think we should prevent it and then add an "auto" mode later on